util/host-utils.c

```diff
@@ -266,3 +266,154 @@ void ulshift(uint64_t *plow, uint64_t *phigh, int32_t shift, bool *overflow)
         *plow = *plow << shift;
     }
 }
+/*
+ * Unsigned 256-by-128 division.
+ * Returns the remainder via r.
+ * Returns lower 128 bit of quotient.
+ */
+static Int128 udiv256_qrnnd(Int128 *r, Int128 n0, Int128 n1, Int128 d)
+{
+    Int128 d0, d1, q0, q1, r1, r0, m;
+    uint64_t mp0, mp1;
+
+    d0 = int128_make64(int128_getlo(d));
+    d1 = int128_make64(int128_gethi(d));
+
+    r1 = int128_remu(n1, d1);
+    q1 = int128_divu(n1, d1);
+    mulu64(&mp0, &mp1, int128_getlo(q1), int128_getlo(d0));
+    m = int128_make128(mp0, mp1);
+    r1 = int128_or(int128_lshift(r1, 64), int128_make64(int128_gethi(n0)));
+    if (int128_lt(r1, m)) {
+        q1 = int128_sub(q1, int128_one());
+        r1 = int128_add(r1, d);
+        if (int128_ge(r1, d)) {
+            if (int128_lt(r1, m)) {
+                q1 = int128_sub(q1, int128_one());
+                r1 = int128_add(r1, d);
+            }
+        }
+    }
+    r1 = int128_sub(r1, m);
+
+    r0 = int128_remu(r1, d1);
+    q0 = int128_divu(r1, d1);
+    mulu64(&mp0, &mp1, int128_getlo(q0), int128_getlo(d0));
+    m = int128_make128(mp0, mp1);
+    r0 = int128_or(int128_lshift(r0, 64), int128_make64(int128_getlo(n0)));
+    if (int128_lt(r0, m)) {
+        q0 = int128_sub(q0, int128_one());
+        r0 = int128_add(r0, d);
+        if (int128_ge(r0, d)) {
+            if (int128_lt(r0, m)) {
+                q0 = int128_sub(q0, int128_one());
+                r0 = int128_add(r0, d);
+            }
+        }
+    }
+    r0 = int128_sub(r0, m);
+
+    *r = r0;
+    return int128_or(int128_lshift(q1, 64), q0);
+}
+
+/*
+ * Unsigned 256-by-128 division.
+ * Returns the remainder.
+ * Returns quotient via plow and phigh.
+ * Also returns the remainder via the function return value.
+ */
+Int128 divu256(Int128 *plow, Int128 *phigh, Int128 divisor)
+{
+    Int128 dhi = *phigh;
+    Int128 dlo = *plow;
+    Int128 rem, dhighest;
+    int sh;
+
+    if (int128_eq(divisor, int128_zero()) || int128_eq(dhi, int128_zero())) {
+        *plow  = int128_divu(dlo, divisor);
+        *phigh = int128_zero();
+        return int128_remu(dlo, divisor);
```

Comment:
And same comment as before, about maybe exposint a divrem-like interface from int128.h to avoid performing the division twice when need both int128_divu and int128_remu to be called